skbio/tree/_nj.py

```diff
@@ -197,12 +199,19 @@ def _compute_collapsed_dm(dm, i, j, disallow_negative_branch_length,
     out_ids = [new_node_id]
     out_ids.extend([e for e in dm.ids if e not in (i, j)])
     result = np.zeros((out_n, out_n))
-    for idx1, out_id1 in enumerate(out_ids[1:]):
-        result[0, idx1 + 1] = result[idx1 + 1, 0] = _otu_to_new_node(
-            dm, i, j, out_id1, disallow_negative_branch_length)
-        for idx2, out_id2 in enumerate(out_ids[1:idx1+1]):
-            result[idx1+1, idx2+1] = result[idx2+1, idx1+1] = \
-                dm[out_id1, out_id2]
+    # pre-populate the result array with known distances
+    ij_indexes = [dm.index(i), dm.index(j)]
+    result[1:, 1:] = np.delete(np.delete(dm.data, ij_indexes, axis=0),
+                               ij_indexes, axis=1)
+    # calculate the new distances from the current DistanceMatrix
+    k_to_u = 0.5 * (dm[i] + dm[j] - dm[i, j])
+    # set negative branches to 0 if specified
+    if disallow_negative_branch_length:
+        k_to_u[k_to_u < 0] = 0
+    # drop nodes being joined
+    k_to_u = np.delete(k_to_u, ij_indexes)
+    # assign the distances to the result array
+    result[0] = result[:, 0] = np.array([0] + k_to_u.data.tolist())
```

Comment:
I think this can be expressed as the following, and avoid the `tolist` call which is relatively expensive:

```python
result[0] = result[:, 0] = np.hstack([0], k_to_u])
```